### ISSUE TITLE
Rancher does not load images behind a proxy (Catalog issues)

### ISSUE THREAD

**Issue by tombmurphy** [rancher/rancher#8335]
With a simple nginx reverse proxy configured on port 80, the rancher management loads ok except for some images which do not load, for example in the Catalog or when choosing a Scheduler when you create an environment.

I think that the doc for running Rancher behind a proxy needs to be more detailed to address issues and correct usage of `no_proxy` directive.

**Rancher Versions:**
v1.5.3

**Docker Version:**
```
$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.6
Storage Driver: btrfs
 Build Version: Btrfs v4.4
 Library Version: 101
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-64-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.952 GiB
Name: rancher-server01
ID: xxxx
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://proxy.company.com:3128
Https Proxy: http://proxy.company.com:3128
No Proxy: localhost,127.0.0.1,.company.com,.company.local
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Setup Details**

Nginx config:

```
upstream rancher {
    server x.x.x.x:8080;
}
server {
	listen 80;

	root /var/www/html;

	server_name rancher.company.com;

	location / {
		proxy_pass http://rancher;
		proxy_set_header Host $host;
        	proxy_set_header X-Forwarded-Proto $scheme;
        	proxy_set_header X-Forwarded-Port $server_port;
        	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        	proxy_http_version 1.1;
        	proxy_set_header Upgrade $http_upgrade;
        	proxy_set_header Connection "upgrade";
        	# This allows the ability for the execute shell window to remain open for up to 15 minutes. Without this parameter, the default is 1 minute and will automatically close.
        	proxy_read_timeout 900s;
	}
}
```

Rancher setup:

```
sudo docker run -d \
    --restart=unless-stopped \
    -p 8080:8080 \
    -e http_proxy=http://proxy.company.com:3128 \
    -e https_proxy=http://proxy.company.com:3128 \
    -e HTTP_PROXY=http://proxy.company.com:3128 \
    -e HTTPS_PROXY=http://proxy.company.com:3128 \
    -e no_proxy="localhost,127.0.0.1,.company.com,.company.local" \
    -e NO_PROXY="localhost,127.0.0.1,.company.com,.company.local" \
    rancher/server:v1.5.3 \
    --db-host ${RANCHER_DB_HOST} \
    --db-port 3306 \
    --db-user ${RANCHER_DB_USERNAME} \
    --db-pass ${RANCHER_DB_PASSWORD} \
    --db-name ${RANCHER_DB_NAME}
```

**Environment Type**
Cattle

**Steps to Reproduce:**
Browse to the Catalog

**Results:** 
Images do not load, looks like a bug ignoring the `no_proxy` directive.

Log:
```
2017-03-29 04:10:20,313 ERROR [:] [] [] [] [qtp859417998-17] [i.g.i.g.r.handler.ExceptionHandler  ] Exception in API for request [http://rancher.company.com/v1/proxy/http://localhost:8088/v1-catalog/templates/library:infra*nfs]. Error id: [d71cd55c-666b-44c6-9300-754ef008bdaa]. org.apache.http.conn.HttpHostConnectException: Connect to localhost:8088 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused
```

**Comment by ibuildthecloud** [comment]
`failed: Connection refused` doesn't seem like it's a proxy issue.  Can you do a `docker exec` into the container and run
```bash
env -i curl http://localhost:8088
```
If all is swell you should get the below output

```
{"data":[{"actions":{},"id":"v1-catalog","links":{"self":"http://localhost:8088/v1-catalog"},"type":"apiVersion"}],"links":{"latest":"http://localhost:8088/v1-catalog","self":"http://localhost:8088/"},"resourceType":"apiVersion","type":"collection"}
```

I more suspect that catalog service isn't running.

**Comment by tombmurphy** [comment]
Thanks @ibuildthecloud I confirm I get exactly the same output.

All images have the broken image sign but if I open the image in a new tab, it opens. I get all the errors in the log though.

**Comment by tombmurphy** [comment]
I'd like to point out that it's not just the Catalog that doesn't load the images, but also when you choose a scheduler when adding a new host. Basically any image.

**Comment by tombmurphy** [comment]
Even if I access Rancher UI directly on port 8080 over VPN, it fails to load half the images in the catalog.
Log:
`2017-04-04 03:41:20,222 ERROR [:] [] [] [] [tp859417998-131] [i.g.i.g.r.handler.ExceptionHandler  ] Exception in API for request [http://rancher-internal:8080/v1/proxy/http://localhost:8088/v1-catalog/templates/community:infra*swarm]. Error id: [072d483d-86d1-456f-ac35-f95b1b253990]. org.apache.http.conn.HttpHostConnectException: Connect to localhost:8088 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused`

**Comment by tombmurphy** [comment]
See attached photos:
![screen shot 2017-04-04 at 1 46 39 pm](https://cloud.githubusercontent.com/assets/2703290/24640805/61828d5e-193d-11e7-9105-a891f01a0cbb.png)

Also when creating a new stack from the Catalog in K8s, no services are created:
![screen shot 2017-04-04 at 1 47 25 pm](https://cloud.githubusercontent.com/assets/2703290/24640815/7ef062e4-193d-11e7-852c-7c6d5b1b69e8.png)

I get a lot of connection refused in logs plus
`time="2017-04-04T03:47:42Z" level=info msg="Starting Catalog Service (port 8088, refresh interval 300 seconds)" `

**Comment by ibuildthecloud** [comment]
The issue here is that each time you access the catalog-service it is just panicing and crashing, so not a proxy issue.  The reason it is dying is due to an odd CGO issue.  We just need to compile the catalog-service to no CGO (as done here https://github.com/rancher/catalog-service/pull/73)

**Comment by tombmurphy** [comment]
Fixed in v1.5.5
